### PR TITLE
Introduce Alternate Toolchain gcc and binutils

### PR DESCRIPTION
**Pull request by klauskiwi**
Hostboot, OCC and potentially other packages are *VERY* sensitive to
GCC+binutils combinations (specially due to the use of libbfd to feed
the custom linker/loader script). That means that the default buildroot
toolchain can't be easily upgraded without risking hard-to-debug build
and runtime issues that might be caused by the change of the compiler +
binutils combination.

Address this by creating (yet another) GCC + Binutils host packages pair
that can be used, as an alternative to the one provided by buildroot.

Doing so would allow such packages to stay (indefinitely) using their
choice of gcc+binutils, while still allowing buildroot (and the Skiroot
environment) to evolve and use newer versions / releases.

Signed-off-by: Klaus Heinrich Kiwi <klaus@linux.vnet.ibm.com>

### CODE REVIEW

File: openpower/configs/zz_defconfig [hunk 1]
```diff
@@ -46,6 +46,7 @@ BR2_TARGET_ROOTFS_CPIO_XZ=y
 BR2_OPENPOWER_PLATFORM=y
 BR2_OPENPOWER_POWER9=y
 # BR2_PACKAGE_PPE42_TOOLCHAIN is not set
+BR2_PACKAGE_ALTERNATE_GCC=y
```
Comment (ibmzach) [inline openpower/configs/zz_defconfig:49]: Are we going to set this variable for P10 configs too?

File: openpower/package/alternate-binutils/alternate-binutils.mk [hunk 2]
```diff
@@ -0,0 +1,56 @@
+################################################################################
+#
+# alternate-binutils
```
Comment (ibmzach) [inline openpower/package/alternate-binutils/alternate-binutils.mk:3]: assuming all of this is identical to the current "primary" toolchain

File: openpower/package/alternate-gcc/Config.in [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+		help
+		  Build gcc for an Alternate Toolchain, for use in
+		  certain packages.
+
+
+if BR2_PACKAGE_ALTERNATE_GCC
+
+config BR2_ALTERNATE_GCC_VERSION
+		string "Alternate gcc version"
+		default "6.5.0"
```
Comment (ibmzach) [inline openpower/package/alternate-gcc/Config.in:13]: is 6.5.0 what we've been using? I feel like we've had GCC 8 at some point

File: openpower/package/alternate-gcc/alternate-gcc.mk [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+define HOST_ALTERNATE_GCC_CONFIGURE_SYMLINK
+	mkdir -p $(@D)/build
+	ln -sf ../configure $(@D)/build/configure
+endef
+
+HOST_ALTERNATE_GCC_CONF_OPTS += \
+	$(call qstrip,$(BR2_ALTERNATE_GCC_EXTRA_CONFIG_OPTIONS))
+
+define  HOST_ALTERNATE_GCC_CONFIGURE_CMDS
```
Comment (ibmzach) [inline openpower/package/alternate-gcc/alternate-gcc.mk:36]: assuming this is all the same

File: openpower/configs/zz_defconfig (hunk 1 above)
Comment (klauskiwi) [reply openpower/configs/zz_defconfig:49]: we could if necessary... I believe we should strive to use buildroot's gcc + binutils the best we can, but if for some reason we need to update buildroot (say to fix security holes or bugs that are only fixed in more modern versions), we could set this setting, as well as the ALTERNATE_*_VERSION to keep hostboot and occ behind if necessary

File: openpower/package/alternate-gcc/Config.in (hunk 3 above)
Comment (klauskiwi) [reply openpower/package/alternate-gcc/Config.in:13]: We were using 6.5.0 from Power8 builds, what the original PR #3564 was trying to solve.. I'll wait for the CI to wrap-up for us to have a baseline and will push a change to get it back to GCC 8.

File: openpower/package/alternate-gcc/alternate-gcc.mk (hunk 4 above)
Comment (klauskiwi) [reply openpower/package/alternate-gcc/alternate-gcc.mk:36]: I tried to make it as close to buildroot's as "necessary" - problem is that they are pretty spread out, and of course, possibly changing with each buildroot release.. see references at https://github.com/open-power/buildroot/blob/a84aaaeacbd1a69053cde2e50f26250e9328de1e/package/gcc/gcc.mk#L73 and https://github.com/open-power/buildroot/blob/a84aaaeacbd1a69053cde2e50f26250e9328de1e/package/gcc/gcc-final/gcc-final.mk#L62

We can probably make the conf command completely static and mimic the buildroot's command verbatin.. if we think this is necessary

File: openpower/package/alternate-binutils/alternate-binutils.mk (hunk 2 above)
Comment (klauskiwi) [reply openpower/package/alternate-binutils/alternate-binutils.mk:3]: seem my other comment

File: openpower/package/alternate-gcc/alternate-gcc.mk (hunk 4 above)
Comment (klauskiwi) [reply openpower/package/alternate-gcc/alternate-gcc.mk:36]: there's also the possibility of adjusting per-config options using `BR2_ALTERNATE_GCC_EXTRA_CONFIG_OPTIONS`

File: openpower/package/alternate-gcc/Config.in (hunk 3 above)
Comment (klauskiwi) [reply openpower/package/alternate-gcc/Config.in]: moved to 8.4.0

**Comment by klauskiwi** [comment]
@dcrowell77 @zane131 should we merge this and monitor CI for the next few weeks? I guess we could revert it if anything...

**Review by ibmzach — COMMENTED** [review]
sounds fine

**Review by ibmzach — APPROVED** [review]

**Comment by dcrowell77** [comment]
Sounds okay to me

**Pull request merged**